framework/src/common/logger.rs

```diff
@@ -0,0 +1,117 @@
+use super::*;
+use failure::Error;
+use fern::colors::{Color, ColoredLevelConfig};
+use fern::Dispatch;
+use itertools::Itertools;
+use std::env;
+use std::io::{stderr, stdout};
+
+/// Logging extension exposed for other libraries to vary additional loggers
+/// upon initialization.
+pub struct LogExt {
+    pub nf: String,
+    pub log_level: log::LevelFilter,
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub enum ErrorLogFormat {
+    Multiline,
+    SingleLine,
+    SingleLineWithoutBacktrace,
+
+    #[doc(hidden)]
+    __Nonexhaustive,
+}
+
+#[macro_export]
+macro_rules! err {
```

Comment:
I think all this can be done simpler as a custom implementation of the `Display` and `Debug` traits on `NetBricksError`. Then you can log any error with `error!("{}", e)` or `warn!("{:?}", e)`. Also no confusion that `error!` and `err!` are in fact two different macros but look almost the same.

Reply:
removed,as backtrace is part of debug calls to `error!` for example.